### PR TITLE
add code fix convert to mapped object type

### LINKED ISSUE #24220: Quick fix for 'unions can't be used in index signatures, use a mapped object type instead'
The following code:

```ts
type K = "foo" | "bar";

interface SomeType {
    [prop: K]: any;
}
```

Gives this error message:

```
An index signature parameter type cannot be a union type. Consider using a mapped object type instead.
```

Nobody knows what mapped object types are, so let's give them a quick fix that

* Switches the index signature to a mapped type
* Moves other members to a separate object type that gets combined with an intersection type
* Changes the containing object type to a type alias if the containing object type is an interface
* Intersects with all the `extends` clauses if the containing object type is an interface and has any `extends` clauses

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #24220

### CODE REVIEW

**Comment by Kingwl** [comment]
some formatting issue with #24312

**Comment by mhegazy** [comment]
@andy-ms can you please review this change.

File: src/services/codefixes/convertToMappedObjectType.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    function isIndexSignatureParameterName(node: Node): node is Identifier {
+        return node && node.parent && node.parent.parent && node.parent.parent.parent &&
+            isIdentifier(node) && isParameter(node.parent) && isIndexSignatureDeclaration(node.parent.parent) && isFixableDeclaration(node.parent.parent.parent);
+    }
+
+    function getFixableSignatureAtPosition(sourceFile: SourceFile, pos: number): Info | undefined {
+        const token = getTokenAtPosition(sourceFile, pos, /*includeJsDocComment*/ false);
+        if (!isIndexSignatureParameterName(token)) return undefined;
```
Comment (andy-ms) [inline src/services/codefixes/convertToMappedObjectType.ts]: Now that Microsoft/vscode#45785 is fixed, might as well just assert these things as you go, which could save you some casts.
The only time this shouldn't be fixable is in a class, right? So in `getCodeActions`, all we should need to determine whether it's fixable is whether `token.parent.parent.parent` is a class?

Comment (Kingwl) [reply src/services/codefixes/convertToMappedObjectType.ts]: thanks, in my memory, i have not to use assert or cast 😬

Comment (Kingwl) [reply src/services/codefixes/convertToMappedObjectType.ts]: after thinking and search, This may exceed my expectations
what should i do?😢

**Comment by andy-ms** [comment]
Thanks!

**Comment by mhegazy** [comment]
No, this will be available in TypeScript 3.0 in two month, or in `typescript@next` coming out later tonight.

**Pull request merged**